### ISSUE TITLE
autoBind doesn't work when used in conjunction with higher-order functions

### ISSUE THREAD

**Issue by ffxsam** [cassiozen/React-autobind#3]
I'm using a package called [Radium](https://www.npmjs.com/package/radium), which is used as a higher-order function:

```
export default Radium(AttachmentCreator)
```

As it turns out, `autoBind` doesn't really like being used with higher-order functions. This is my code:

```
class AttachmentCreator extends Component {
  constructor(props) {
    super(props);
    autoBind(this);

    this.state = {
      chosenType: 'FILE',
      open: false,
    };
  }

  // ...
```

The strange thing is, in a development environment, it works fine. But in production, if a class method tries to refer to `this`, it doesn't refer to the React element.

**Comment by andrewbranch** [comment]
+1, I'm using react-css-modules and have the same problem. `this` refers to the wrapper component. (Not sure about the dev vs. production observation.) It does seem counterintuitive... could this be a Babel idiosyncrasy? It’s hard to follow what’s happening in the transpiled code.

**Comment by ffxsam** [comment]
I don't think it has anything to do with babel. I think it's because we're wrapping the React component in something else, and then `autoBind` is confused and starts messing with the wrapper.

**Comment by andrewbranch** [comment]
Ah, I just realized that react-css-modules' `WrappedComponent` uses inheritance, not higher-order components. This explains my issue, and I’m not sure if there’s an elegant solution to implement in this package.

I’m not sure if this would have any adverse side effects, but you could autobind recursively up the prototype chain until the prototype is `ReactComponent` itself to handle multiple inheritance. That seems a little sketchy to me to implement by default in this library, but maybe it could be turned on with an option. Alternatively, I might just do that in my own component's constructor, knowing the component is going to be wrapped by CSSModules.

**Comment by ffxsam** [comment]
Radium uses a higher order function, and autobind doesn't like that either.

**Comment by andrewbranch** [comment]
This is not too bad I guess:

``` js
constructor(props) {
  super(props);
  autobind(Object.getPrototypeOf(this));
}
```

**Comment by ffxsam** [comment]
I'll give that a shot, thanks!

**Comment by ffxsam** [comment]
@andrewbranch Well, that kinda works. In a dev environment, it fails, but in production it works. Very strange.

I'm just gonna fork this project and make a new package, and fix this issue—and possibly add some other features.

**Comment by cassiozen** [comment]
Hi @ffxsam, 

I'm sorry it took me this long to see this issue. I'll also take a look at this issue over the next few days. Anyway, are you interested in talking about being a mantainer of this project?

**Comment by cassiozen** [comment]
@ffxsam, sorry it took me this long to see this issue. I'll also take a look at this issue over the next days.

**Comment by andrewbranch** [comment]
For the record, I don’t have any behavior disparity between dev and production. @ffxsam I think you must have something strange going on with some build dependency or something.

**Comment by ffxsam** [comment]
@andrewbranch Probably because I'm using Meteor, I'm guessing.

@cassiozen Love to! If you could make me a maintainer of the GitHub repo and grant permissions for me to publish to npm, that would be cool (my npmjs.com username is ffxsam).

**Comment by ffxsam** [comment]
@cassiozen Yes.. please. :)

**Comment by cif** [comment]
+1  Any news on this?

**Comment by ffxsam** [comment]
@cif Using third party auto-binding libraries is no longer necessary. If you're using a babel, you can declare React methods like so:

``` js
  myMethod = (args) => {
  };
```

and it will be auto-bound for you.

**Comment by andrewbranch** [comment]
@cif I've been using `autobind(Object.getPrototypeOf(this))` pretty consistently in all sorts of environments. It makes sense while not immediately intuitive. I prefer real methods with autobinding to arrow function property initializers as @ffxsam recommended, for mostly aesthetic reasons. To split hairs a little bit in the hopes of avoiding confusion for someone else down the road, `myMethod = args => {};` is not auto-bound; rather, it inherits/binds no context at all, using the outer closure’s context as its `this`. ES7 property initializers like that actually happen in the constructor, so it would be transpiled down to ES6 something like this:

``` js
constructor(...constructorArgs) {
  super(...constructorArgs);
  this.myMethod = args => {
    // `this` just refer’s to the `this` in the constructor body, because arrow functions
  };
}
```

So the effect is the same as autobinding (your `this` is correct, yay), by slightly different means. Like I said, splitting hairs, but in some cases it can be an important distinction. At any rate, what @ffxsam is a good solution that a lot of people use and enjoy.

**Comment by cif** [comment]
@ffxsam Does anyone not use babel now days? ; ) Where in your "class" are you declaring that method so that `this.myMethod` is bound correctly? 

Ideally, I don't want handler methods declared in the constructor (thanks @andrewbranch for suggestion though) nor any other method really, especially `render()` as I've seen from bad habits since this re-creates that method every time props or state changed.

```
class Testing extends React.Component {
  handleClick = (event) => {  // <-- syntax error
    alert(this);
  }
  render() {
    return (<a onClick={this.handleClick}>hello</a>);
  }
}
```

```
class Testing extends React.Component {
  handleClick(event) {
    alert(this); // <-- this = undefined
  }
  render() {
    return (<a onClick={this.handleClick}>hello</a>);
  }
}
```

**Comment by cif** [comment]
Dudes. Sorry. I posted this in the wrong autobind repo (looking for decorator). 

That said, it's apparently an issue for them as well, and decorators are just fancy HOFs.

**Comment by ffxsam** [comment]
@cif Your code looks good (the first block). Make sure you have the stage-1 babel preset installed to be able to use the syntax I suggested.

**Comment by cif** [comment]
@ffxsam beauty, thanks! works like a charm.

**Comment by dnish** [comment]
@ffxsam how would you solve this issue when binding those functions via babel:

```
class Test {

    constructor() {

    }


    message = (text) => {
        console.log(text);
    }
}


class Test2 extends Test {

    constructor(props) {
        super(props);
    }

    message = (text) => {
        super.message(text);
        console.log(text);
    }
}

let test = new Test2();
test.message("HEY");
```

When calling super, this will end up with the following error message:

    SyntaxError: 'super' outside of function or class

**Comment by ffxsam** [comment]
That's a really strange pattern.. I'm not sure why you would even do that.

**Comment by dnish** [comment]
@ffxsam in a normal class I wouldn't do this, but within a React class I want to "autobind" my methods. In this case I have a 2 components (Player and MobilePlayer) where MobilePlayer extends Player. Both share the same methods but MobilePlayer has some additions. There I need to call super to share the method logic, but this won't work with this binding pattern (both methods need to access this.state).

**Comment by ffxsam** [comment]
@dnish I would probably just use higher-order components instead of class inheritance. I've never really seen people use inheritance with React classes like that.

https://facebook.github.io/react/docs/higher-order-components.html
